pkg/reconciler/pipelinerun/resources/workspaces_test.go

```diff
@@ -0,0 +1,166 @@
+			tb.PipelineTask("pt1", "task",
+				tb.PipelineTaskWorkspaceBinding("foo", tb.PipelineTaskWorkspaceBindingWorkspace("named-workspace1")),
+			),
+			tb.PipelineTask("pt2", "task",
+				tb.PipelineTaskWorkspaceBinding("foo", tb.PipelineTaskWorkspaceBindingWorkspace("named-workspace1")),
+			),
+			tb.PipelineTask("pt3", "task",
+				tb.PipelineTaskWorkspaceBinding("foo", tb.PipelineTaskWorkspaceBindingWorkspace("named-workspace1")),
+			),
+		)),
+	}, {
+		description: "multiple variables resolving to different workspace names",
+		in: tb.Pipeline("test-pipeline", "foo", tb.PipelineSpec(
+			tb.PipelineWorkspaceDeclaration("named-workspace1", "named-workspace2"),
+			tb.PipelineTask("pt1", "task",
+				tb.PipelineTaskWorkspaceBinding("foo", tb.PipelineTaskWorkspaceBindingWorkspace("named-workspace1")),
+				tb.PipelineTaskWorkspaceBinding("bar", tb.PipelineTaskWorkspaceBindingWorkspace("named-workspace2")),
+			),
+			tb.PipelineTask("pt2", "task",
+				tb.PipelineTaskWorkspaceBinding("foo", tb.PipelineTaskWorkspaceBindingFrom("pt1", "foo")),
+			),
+			tb.PipelineTask("pt3", "task",
+				tb.PipelineTaskWorkspaceBinding("foo", tb.PipelineTaskWorkspaceBindingFrom("pt1", "bar")),
+			),
+			tb.PipelineTask("pt4", "task",
+				tb.PipelineTaskWorkspaceBinding("foo", tb.PipelineTaskWorkspaceBindingFrom("pt2", "foo")),
+			),
+			tb.PipelineTask("pt5", "task",
+				tb.PipelineTaskWorkspaceBinding("foo", tb.PipelineTaskWorkspaceBindingFrom("pt3", "foo")),
+			),
+		)),
+		expected: tb.Pipeline("test-pipeline", "foo", tb.PipelineSpec(
+			tb.PipelineWorkspaceDeclaration("named-workspace1", "named-workspace2"),
+			tb.PipelineTask("pt1", "task",
+				tb.PipelineTaskWorkspaceBinding("foo", tb.PipelineTaskWorkspaceBindingWorkspace("named-workspace1")),
+				tb.PipelineTaskWorkspaceBinding("bar", tb.PipelineTaskWorkspaceBindingWorkspace("named-workspace2")),
+			),
+			tb.PipelineTask("pt2", "task",
+				tb.PipelineTaskWorkspaceBinding("foo", tb.PipelineTaskWorkspaceBindingWorkspace("named-workspace1")),
+			),
+			tb.PipelineTask("pt3", "task",
+				tb.PipelineTaskWorkspaceBinding("foo", tb.PipelineTaskWorkspaceBindingWorkspace("named-workspace2")),
+			),
+			tb.PipelineTask("pt4", "task",
+				tb.PipelineTaskWorkspaceBinding("foo", tb.PipelineTaskWorkspaceBindingWorkspace("named-workspace1")),
+			),
+			tb.PipelineTask("pt5", "task",
+				tb.PipelineTaskWorkspaceBinding("foo", tb.PipelineTaskWorkspaceBindingWorkspace("named-workspace2")),
+			),
+		)),
+	}} {
+		t.Run(tc.description, func(t *testing.T) {
+			result, err := ResolveWorkspaces(tc.in.Spec.DeepCopy())
+			if err != nil {
+				t.Errorf("error resolving workspace: %v", err)
+			}
+			if d := cmp.Diff(&tc.expected.Spec, result); d != "" {
+				t.Errorf("ResolveWorkspaces() got diff %s", d)
+			}
+		})
+	}
+}
+
+func TestResolveWorkspaces_Invalid(t *testing.T) {
+	for _, tc := range []struct {
+		description  string
+		in           *v1alpha1.Pipeline
+		errorMessage string
```

Comment:
this is mostly personal preference but i usually avoid testing against error messages - feels like there's not a lot of value if they change a bit and then the test case needs to change too (similar feelings for log messages)

but im guessing the reason might actually be that you want to make sure the error happened for hte right reason? if so, then i have 2 alternatives to suggest:
1. return typed errors and assert agains the type
2. break up the funcitonality into multiple functions, which each fail or not for their own reasons